Pot draft 2 (#8)
subspace/subnomicon: docs/consensus/pot.md

---
title: Proof-of-Time
sidebar_position: 5
description: Proof-of-Time consensus component
keywords:
    - Consensus
    - randomness
    - challenge
---
Dilithium is secured by a Proof-of-Time (PoT) component to introduce unpredictability and deter long-range attacks. 

## Availability and Unpredictability

> The permissionless Proof-of-Work (PoW) introduced by Bitcoin remains the most robust method for achieving consistent availability and unpredictability in a decentralized system. In longest-chain PoW, a new block can be appended to the tip of the chain by a miner who solves a cryptographic hash challenge. However, mining requires massive energy expenditure to brute-force hash solutions, which has led to the introduction of various alternative consensus mechanisms. Energy-efficient proof systems, including Proof-of-Stake and Proof-of-Capacity, thrive for certain important features of PoW, like dynamic availability. Bitcoin has been continuously available for over a decade despite an always varying hashrate due to miners joining and leaving the network. 

Reviewer: >introduced by Bitcoin

Maybe "used in Bitcoin", as it was introduced prior to Bitcoin.

Reviewer: Just making sure that you decided to keep it as it was?

Author: Changed to used now. I was sure I committed it 🤦‍♀️

> The permissionless Proof-of-Work (PoW) introduced by Bitcoin remains the most robust method for achieving consistent availability and unpredictability in a decentralized system. In longest-chain PoW, a new block can be appended to the tip of the chain by a miner who solves a cryptographic hash challenge. However, mining requires massive energy expenditure to brute-force hash solutions, which has led to the introduction of various alternative consensus mechanisms. Energy-efficient proof systems, including Proof-of-Stake and Proof-of-Capacity, thrive for certain important features of PoW, like dynamic availability. Bitcoin has been continuously available for over a decade despite an always varying hashrate due to miners joining and leaving the network. 

Reviewer: >achieving consistent availability and unpredictability

If mention these terms, I think we really need to explain what they mean. It can be elsewhere, like the Terminology page, but then we should provide a link.

> Subspace's consensus obtains a farming dynamic that mimics the random time interval of Bitcoin's mining dynamic while only expending a small constant amount of electricity. This is achieved through a Proof-of-Time for block proposal lottery based on the fundamental paper [PoSAT: Proof-of-Work Availability and Unpredictability, without the Work](https://arxiv.org/abs/2010.08154) by Soubhik Deb, Sreeram Kannan and David Tse. It ensures the fairness of the farming process to all participants through complete unpredictability of who will get to propose a block next. This unpredictability is at the same level as PoW protocols and is stronger than existing PoStake protocols using Verifiable Random Functions.

Reviewer: > This is achieved through a Proof-of-Time for block proposal lottery based on the fundamental paper [PoSAT: Proof-of-Work Availability and Unpredictability, without the Work](https://arxiv.org/abs/2010.08154)

Just to note that in subspace the PoT design differs from this paper (global vs. local PoT). Also to note that it's not the first work that introduced this concept.

>than existing PoStake protocols using Verifiable Random Functions.

Proof-of-Stake? or write above "Proof-of-Stake (PoStake)", as with PoW. Though there's nothing particular about PoStake here, just VRF-based protocols.
I will also not capitalised "Verifiable Random Functions".

Author: I think we can still write "based" on the paper, but I removed the "fundamental" part

## Long-range attack

Unlike in Proof-of-Work, the process of block production in Proof-of-Stake and Proof-of-Capacity-based
blockchains is not physically constrained. This opens the door to a _long-range attack_.

> A long-range attack refers to an adversary controlling enough resources at some point of chain life to rewrite a significant portion of the chain history. In PoW protocols like Bitcoin, this requires controlling over 50% of the total network hashrate for a sustained period of time which is infeasible in practice. However, long-range attacks remain a serious threat in alternative consensus protocols that do not rely on proof-of-work.

Reviewer: 1. We need to make the point that it's a vulnerability because generating a fork all the way to the tip is easy (see [here](https://github.com/subspace/subnomicon/pull/5/files#diff-6e7fdf1a991d122bf2209340f35997bcb63fb872660a132c0733c3c3121dd0aaR18)).
2. In Bitcoin this attack is prevented because it is fundamentally hard to generate a fork in a short amount of time. This is the reason we mention the physical constraint above.

Author: Is it better now?

Reviewer: We are still talking about Bitcoin at two different places in this section, once about >50% hashrate and once about the physical constraint.

> and that chain will always be behind the current longest chain

Not always, as with >50% the fork will eventually catch up.

Since with dishonest 51% there are no security guarantees (long or short), I don't see a real reason to mention it here, but if you decide to do so, I advise that we concentrate all things Bitcoin in one place. Note that even with a 51% attack, there's a difference as it will take a substantial time to complete a longer chain.

Suppose in the first year of the blockchain's operation, all farmers were honest and have collectively pledged 100TB of storage. Now, suppose by the second year the total storage pledged reached 1PB, out of which an adversary has dedicated 200TB worth of storage. At no point does an adversary control more than 20% of storage, which is way less than a majority. However, using his 200TB, an adversary could rewrite the past year's history by participating in all past lotteries to win blocks back to the genesis and then grow a chain instantaneously from the genesis to surpass the current longest chain. This is possible because the adversary's resources are enough to win a disproportionately large number of past lotteries compared to his share of total storage. Such a long-range rewrite seriously threatens the security and immutability of blockchain history.
On the other hand, in Bitcoin it takes a long time to mine such a chain from the past and that chain will always be behind the current longest chain. Thus, PoW enforces an arrow of time, meaning nodes cannot “go back in time” to mine blocks. This property is key to tolerating a fully dynamic honest and adversarial participation.

> ## Arrow of Time

Reviewer: I'm not sure this title is proper. Here we explain how Subspace prevents the above.

Author: How about Solution? Or Subspace Arrow-of-Time? Or just merge with prev section

Reviewer: Subspace solution or approach?

> Subspace's Proof-of-Time component addresses long-range attacks by enforcing an arrow of time similar to PoW protocols. PoT guarantees a certain amount of wall-clock time must elapse between block proposals, preventing an adversary from rewriting history by "going back in time". Similar to PoW, Proof-of-Time is constrained physically, however it is not parallelizable (technically, it is proof of _sequential_ work). We prevent the aforementioned attack by integrating the blockchain with a Proof-of-Time process. The attacker can not immediately generate a years-long fork on the spot even with faster hardware.

Reviewer: ```suggestion
Subspace's Proof-of-Time component addresses long-range attacks by enforcing an arrow of time similar to PoW protocols. PoT guarantees that a certain amount of wall-clock time must elapse between block proposals, preventing an adversary from rewriting history by "going back in time". Similar to PoW, Proof-of-Time is constrained physically, however it is not parallelizable (technically, it is proof of _sequential_ work). We prevent the aforementioned attack by integrating the blockchain with a Proof-of-Time process. The attacker can not immediately generate a years-long fork on the spot even with faster hardware.

The elapsed time guarantee is achieved by iterative evaluation of an inherently sequential function. The output of such a function is unpredictable and is used to build a randomness beacon for block challenges. 

## Timekeeping

> For the task of running the time-chain Subspace introduces a new role for nodes called Timekeepers. Timekeepers are responsible for evaluating the delay function and announcing the outputs to other nodes. Anyone can become a Timekeeper as long as they have a sufficiently powerful CPU to be able to evaluate the delay function within the target time slot duration of 1 second. A timekeeper can also be a farmer and participate in block production or an operator executing computation on a domain. 

Reviewer: ```suggestion
For the task of running the time-chain, Subspace introduces a new role for nodes called Timekeepers. Timekeepers are responsible for evaluating the delay function and announcing the outputs to other nodes. Anyone can become a Timekeeper as long as they have a sufficiently powerful CPU to be able to evaluate the delay function within the target time slot duration of 1 second. A timekeeper can also be a farmer and participate in block production or an operator executing computation on a domain.

> A single timekeeper is sufficient for the security of the protocol, but for robustness and decentralization there will be multiple timekeepers running in parallel. We invite interested parties to run timekeeper component on their nodes to ensure the security and decentralization of the protocol. Domain operators may be more suited for the task since they likely already have powerful hardware.

Reviewer: >A single timekeeper is sufficient for the security of the protocol, 

Does it need to be honest?

>but for robustness and decentralization there will be multiple timekeepers running in parallel.

I'm not sure that I get this, how can we guarantee that there will be multiple ones?

Author: Right, it needs to be honest. In worst case, we will run multiple in-house but we also have a few eager community members. So we can't "guarantee" I guess but it will likely be the case

Reviewer: If subspace runs several timekeepers, from security perspective it is really as if only one is being run -- it definitely don't achieve decentralisation. I just think we should soften the language here.

Author: For example?

Reviewer: "there will be" --> "there should be"

The fact that we have a sequence of random values coming from the Proof-of-Time evaluation allows us to use it
as a source of randomness for block production. Since we target a block challenge every second, we can set the delay function evaluation to output a proof every second. Then for every time slot, timekeepers will evaluate the delay function for a set number of iterations to generate fresh global randomness. They then announce the output to the network, which is used by farmers to determine the next block proposer.

<!-- ![ProofOfTimeChallenges](../../../src/Images/PoTChallenges.png) -->

Farmers receive fresh randomness from timekeepers, verify it and scan their plots to see if they contain any chunks of history close enough to the challenge threshold to claim the block. Farmers with the correct chunks provide a proof-of-space for those, propose a block and earn rewards. The randomness is revealed a few slots in advance to ensure every farmer on the network has had enough time to receive it, scan their plots and submit the proof-of-space if they win. The farmers include PoT outputs in the block header, and the PoT chain is persisted in the consensus chain in this way.

> Every 50 blocks, entropy from the consensus chain is injected back into the PoT chain. The injection takes the farming solution and PoT output from a deep consensus block header as the new input for the delay function. Injection also prevents an adversary from simulating a PoT chain without also having to simulate a consensus chain fork. Forks in the consensus chain will result in a different sequence of PoT outputs. Hence, an attacker that forks the chain in some historical point, will have to physically run the PoT algorithm. 

Reviewer: > Injection also prevents an adversary from simulating a PoT chain without also having to simulate a consensus chain fork

It's the other way around: it prevents simulating a new consensus fork with the old pot chain

---
title: Proof-of-Time
sidebar_position: 5
description: Proof-of-Time consensus component
keywords:
    - Consensus
    - randomness
    - challenge
---

> In addition to Proof-of-Space component described in previous section, Dilithium is secured by a Proof-of-Time (PoT) component. The chosen PoT algorithm is sequential AES, tuned for 1 proof per second. PoT is a measure against long-range attacks, and addresses unpredictability and dynamic availability issues. 

Reviewer: ```suggestion
In addition to the Proof-of-Space component described in the previous section, Dilithium is secured by a Proof-of-Time (PoT) component. The chosen PoT algorithm is sequential AES, tuned for 1 proof per second. PoT is a measure against long-range attacks, and addresses unpredictability and dynamic availability issues.

> In addition to Proof-of-Space component described in previous section, Dilithium is secured by a Proof-of-Time (PoT) component. The chosen PoT algorithm is sequential AES, tuned for 1 proof per second. PoT is a measure against long-range attacks, and addresses unpredictability and dynamic availability issues. 

Reviewer: ```suggestion
In addition to the Proof-of-Space component described in the previous section, Dilithium is secured by a Proof-of-Time (PoT) component. The chosen PoT algorithm is sequential AES, tuned for 1 second per proof. PoT is a measure against long-range attacks, and addresses unpredictability and dynamic availability issues.

```
The extra  line is on purpose, to make the next block of text in a new paragraph

The permissionless Proof-of-Work (PoW) used by Bitcoin remains the most robust consensus method in decentralized systems. However, mining requires massive energy expenditure to brute-force hash solutions, which has led to the introduction of various alternative consensus mechanisms. When transitioning to energy-efficient proof systems, like Proof-of-Stake and Proof-of-Space, however, several issues arise. 

> ## Long-range attack

Reviewer: ```suggestion

# Design challenges
## Long-range attack

Unlike in Proof-of-Work, the process of block production in Proof-of-Stake and Proof-of-Space-based

> blockchains is not physically constrained. This makes such protocols vulnerable to a _long-range attack_, when an attacker can produce, very quickly a chain all the way from genesis to the current time, and this chain can potentially be heavier than the current "canonical" chain.

Reviewer: ```suggestion
blockchains is not physically constrained. This makes such protocols vulnerable to a _long-range attack_, where an attacker can produce, very quickly, an alternative chain all the way to the current time, and this chain can potentially be heavier than the current "canonical" chain.
```
removed "from genesis", because it is not necessary in the attack.

<!-- ![LongRangeAttack](../../../src/Images/Long-Range_Attack.png) -->

> A long-range attack refers to an adversary controlling enough resources at some point of chain life to rewrite a significant portion of the chain history. In PoW protocols like Bitcoin, this requires controlling over 50% of the total network hashrate for a sustained period of time which is infeasible in practice. However, long-range attacks remain a serious threat in alternative consensus protocols that do not rely on Proof-of-Work.

Reviewer: ```suggestion
```
This just repeats the previous paragraph.
If you decide to keep this text, please make a new paragraph for the following text.

Author: Rephrased it a bit and added new line

Reviewer: I still prefer not to mention that Bitcoin needs 50% here, because in general generating an alternative chain in Bitcoin requires a long time, and we write above that a long-range attack is when an alternative chain can be formed very quickly (which is probably just a restriction). We can also change the definition, but I'm happy with the focus on it (otherwise, we should also consider these attacks against PoT).

Ultimately your call.

> Let's show how a Proof-of-Space blockchain can be vulnerable to long-range attacks. An equivalent setup is also true for Proof-of-Stake protocols.

Reviewer: ```suggestion
Let's show how a Proof-of-Space blockchain can be vulnerable to long-range attacks. Similar attack applies to Proof-of-Stake.

Suppose in the first year of the blockchain's operation, all farmers were honest and have collectively pledged 100TB of storage. Now, suppose by the second year the total storage pledged reached 1PB, out of which an adversary has dedicated 200TB worth of storage. At no point does an adversary control more than 20% of storage, which is way less than a majority. However, using his 200TB, an adversary could rewrite the past year's history by participating in all past lotteries to win blocks back to the genesis and then grow a chain instantaneously from the genesis to surpass the current longest chain. This is possible because the adversary's resources are enough to win a disproportionately large number of past lotteries compared to his share of total storage. Such a long-range rewrite seriously threatens the security and immutability of blockchain history.

> On the other hand, in PoW this attack is prevented as it takes a long time to mine an alternative chain from the past. Thus, PoW enforces an arrow of time, meaning nodes cannot “go back in time” to mine blocks. This property is key to tolerating a fully dynamic honest and adversarial participation.

Reviewer: ```suggestion
On the other hand, in PoW this attack is prevented as it takes a long time to mine an alternative chain from the past. Thus, PoW enforces an arrow of time, where it is not practical to “go back in time” (unless the attacker has more than 50% of the current hashrate). This property is key to tolerating a fully dynamic honest and adversarial participation.

## Availability and Unpredictability 

Alternative proof systems, including Proof-of-Stake and Proof-of-Space, strive for certain important features of PoW, like dynamic availability and unpredictability.

Dynamic availability in blockchains refers to the capacity of the system to maintain robust operation in environments where nodes may join or leave dynamically. The permissionless PoW remains the most robust method for achieving consistent availability and unpredictability in a decentralized system. Bitcoin has been continuously available for over a decade despite an always varying hashrate due to miners joining and leaving the network. 

> In this setting, unpredictability refers to the inability to predict who will get to propose the next block. Unpredictability of block proposers ensures the fairness of the farming process to all participants. However, the protocols using verifiable random functions do not achieve this property at the same level as in PoW and suffer from predictability of block challenges.

Reviewer: ```suggestion
In this setting, unpredictability refers to the inability to predict who will get to propose the next block. Unpredictability of block proposers is important for the security and liveness of the network. However, protocols using generic verifiable random functions to elect block proposers usually do not achieve this property at the same level as in PoW, and may suffer from a long predictability window of block challenges.```

We need to be cautious here, as our PoT is also VRF

Author: applied suggestion manually without ``` in the end of sentence

Subspace's Proof-of-Time component addresses long-range attacks by enforcing an arrow of time similar to PoW protocols. PoT guarantees that a certain amount of wall-clock time must elapse between block proposals, preventing an adversary from rewriting history by "going back in time". Similar to PoW, Proof-of-Time is constrained physically, however it is not parallelizable (technically, it is proof of _sequential_ work). We prevent the aforementioned attack by integrating the blockchain with a Proof-of-Time process. The attacker can not immediately generate a years-long fork on the spot even with faster hardware.

> The elapsed time guarantee is achieved by iterative evaluation of an inherently sequential function. The output of such a function is unpredictable and is used to build a randomness beacon for block challenges. 

Reviewer: ```suggestion
Subspace uses Proof-of-Time to a form a separate Proof-of-Time chain. The Proof-of-Space and Proof-of-Time chains are inter-connected, preventing long-range attacks. Additionally, challenges for block farming are based on the PoT outputs, which guarantee that the challenges, hence the block proposers, are not predictable.

Subspace's Proof-of-Time component addresses long-range attacks by enforcing an arrow of time similar to PoW. PoT guarantees that a certain amount of wall-clock time must elapse between block proposals, preventing an adversary from rewriting history by "going back in time". Similar to PoW, Proof-of-Time is constrained physically, however it is not parallelizable (technically, it is proof of _sequential_ work). We prevent the aforementioned attack by integrating the blockchain with a Proof-of-Time process. The attacker can not immediately generate a years-long fork on the spot even with faster hardware.

Subspace's consensus obtains a farming dynamic that mimics the random nature of Bitcoin's mining dynamic while only expending a small constant amount of electricity. This is achieved through a Proof-of-Time-based block challenges for the block proposal lottery, based on the paper [PoSAT: Proof-of-Work Availability and Unpredictability, without the Work](https://arxiv.org/abs/2010.08154) by Soubhik Deb, Sreeram Kannan and David Tse. It ensures the fairness of the farming process to all participants through complete unpredictability of who will get to propose a block next. This unpredictability is at the same level as PoW protocols and is stronger than in the protocols using verifiable random functions.

The elapsed time guarantee is achieved by iterative evaluation of an inherently sequential function. The output of such a function is unpredictable and is used to build a randomness beacon for block challenges.

## Timekeeping

For the task of running the time-chain, Subspace introduces a new role for nodes called Timekeepers. Timekeepers are responsible for evaluating the delay function and announcing the outputs to other nodes. Anyone can become a Timekeeper as long as they have a sufficiently powerful CPU to be able to evaluate the delay function within the target time slot duration of 1 second. A timekeeper can also be a farmer and participate in block production or an operator executing computation on a domain. 
A single honest timekeeper is sufficient for the security of the protocol, but for robustness and decentralization there should be multiple timekeepers running in parallel. We encourage interested participants to run the timekeeper component on their nodes to ensure the security and decentralization of the protocol. Domain operators may be more suited for the task since they likely already have powerful hardware.

> The timekeepers start the Proof-of-Time chain starts at the genesis time of the Subspace consensus chain. The input to the first slot is a random seed which will be publicly announced at launch to ensure equal opportunity. For each subsequent slot, the output of the previous slot serves as the input. By chaining the outputs, the timekeepers enforce sequentiality and prevent skipping ahead in time.

Reviewer: ```suggestion
The Proof-of-Time chain starts at the genesis time of the Subspace consensus chain. The input to the first slot is a random seed which will be publicly announced at launch to ensure equal opportunity. For each subsequent slot, the output of the previous slot serves as the input. By chaining the outputs, the timekeepers enforce sequentiality and prevent skipping ahead in time.
